## ***Exploratory Data Analysis on Global_SuperStore***

### ***Objective***
Conduct a detailed analysis of the Global Superstore Database to uncover valuable insights across multiple business dimensions, such as sales trends, profitability, customer behavior, product performance, and logistics. This project aims to enable data-driven decisions that improve profitability, customer satisfaction, and operational efficiency.

### ***Scope of Analysis***

1.	**Sales Trends and Patterns:** Identify seasonal fluctuations and observe long-term trends in sales across different product categories and regions.

2.	**Profitability Analysis:** Determine profit margins across product categories, customer segments, and regions to pinpoint high-margin areas.

3.	**Customer Segmentation:** Segment customers based on purchasing behavior to determine high-value customer groups and preferences.

4.	**Product Performance:** Assess the performance of products and categories to identify top-sellers and items with underperformance.

5.	**Regional Analysis:** Evaluate sales and profitability by region to identify growth opportunities and regional customer preferences.

6.	**Shipping Dynamics:** Analyze shipping preferences, delivery times, and costs to optimize logistics and enhance customer satisfaction.

7.	**Promotional Effectiveness:** Evaluate the impact of promotions and discounts on sales and profitability.

8.	**Return Analysis:** Investigate return rates and underlying reasons to minimize returns and improve product quality.

### ***Importing Necessary Libraries***
Importing the necessary libraries which are required for our analysis, such as Data Loading, Statistical Analysis, Data Visualizations, Data Transformations, Merge and Joins etc.

Numpy and Pandas are used for Data Manipulation and Numerical Calculations.

Matplotlib and Seaborn are used for Data Visualizations.

In [131]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import mysql.connector

import warnings
warnings.filterwarnings('ignore')

### ***Connecting SQL with Python***

In [132]:
Conn = mysql.connector.connect(
    host='localhost',               # host name
    user='root',                    # the user who has privilege to the database
    password='root',                # password for user
    database='eda_practice',        # database name
)

In [133]:
query = 'select * from global_superstore'

In [134]:
df = pd.read_sql(query, Conn)

In [135]:
df.head(2)

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   

    Customer ID  Customer Name    Segment  Postal Code           City  ...  \
0  AB-100151402  Aaron Bergman   Consumer      73120.0  Oklahoma City  ...   
1     JR-162107  Justin Ritter  Corporate          NaN     Wollongong  ...   

     Category Sub-Category                               Product Name  \
0  Technology       Phones                           Samsung Convoy 3   
1   Furniture       Chairs  Novimex Executive Leather Armchair, Black   

     Sales Quantity Discount    Profit Shipping Cost  Order Priority  Returned  
0   221.98        2      0.0   62.1544         40.77            High      None  
1  3709.40        9      0.1 -288.7650        923.63        Critical      None  

[2 rows x 25 columns]

### ***Understanding the Dataset***
This step of EDA helps in understanding the whole dataset which involves the number of rows and columns, datatypes and missing values present in it.

In [136]:
df.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment  Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer      73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate          NaN     Wollongong   
2     CR-127307      Craig Reiter     Consumer          NaN       Brisbane   
3    KM-1637548  Katherine Murray  Home Office          NaN         Berlin   
4    RH-9495111       Rick Hansen     Consumer          NaN          Dakar   

   ...    Category Sub-Category                               Product Name  \
0  ...  Technology       Phones                           Samsung Convoy 3   
1  ...   Furniture       Chairs  Novimex Executive Leather Armchair, Black   
2  ...  Technology       Phones          Nokia Smart Phone, with Caller ID   
3  ...  Technology       Phones             Motorola Smart Phone, Cordless   
4  ...  Technology      Copiers             Sharp Wireless Fax, High-Speed   

     Sales Quantity Discount    Profit Shipping Cost  Order Priority  Returned  
0   221.98        2      0.0   62.1544         40.77            High      None  
1  3709.40        9      0.1 -288.7650        923.63        Critical      None  
2  5175.17        9      0.1  919.9710        915.49          Medium      None  
3  2892.51        5      0.1  -96.5400        910.16          Medium      None  
4  2832.96        8      0.0  311.5200        903.04        Critical      None  

[5 rows x 25 columns]

The head() function of pandas helps in knowing the top few rows of the dataset which gives a glimpse of the whole dataset.

In [137]:
df.tail()

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
51285   29002   IN-2015-KE1642066-42174 2015-06-19 2015-06-19        Same Day   
51286   34337  US-2014-ZD21925140-41765 2014-05-06 2014-05-10  Standard Class   
51287   31315  CA-2012-ZD21925140-41147 2012-08-26 2012-08-31    Second Class   
51288    9596   MX-2013-RB1979518-41322 2013-02-17 2013-02-21  Standard Class   
51289    6147   MX-2013-MC1810093-41416 2013-05-22 2013-05-26    Second Class   

        Customer ID       Customer Name      Segment  Postal Code  \
51285    KE-1642066     Katrina Edelman    Corporate          NaN   
51286  ZD-219251408  Zuschuss Donatelli     Consumer      37421.0   
51287  ZD-219251404  Zuschuss Donatelli     Consumer      94109.0   
51288    RB-1979518          Ross Baird  Home Office          NaN   
51289    MC-1810093       Mick Crebagga     Consumer          NaN   

                City  ...         Category Sub-Category  \
51285           Kure  ...  Office Supplies    Fasteners   
51286    Chattanooga  ...        Furniture  Furnishings   
51287  San Francisco  ...  Office Supplies          Art   
51288       Valinhos  ...  Office Supplies      Binders   
51289       Tipitapa  ...  Office Supplies        Paper   

                                        Product Name  Sales Quantity Discount  \
51285                  Advantus Thumb Tacks, 12 Pack  65.10        5      0.0   
51286  Eldon Image Series Desk Accessories, Burgundy  16.72        5      0.2   
51287                                     Newell 341   8.56        2      0.0   
51288                        Acco Index Tab, Economy  13.44        2      0.0   
51289        Eaton Computer Printout Paper, 8.5 x 11  61.38        3      0.0   

       Profit Shipping Cost  Order Priority  Returned  
51285  4.5000         1.010          Medium      None  
51286  3.3440         1.930            High      None  
51287  2.4824         1.580            High      None  
51288  2.4000         1.003          Medium      None  
51289  1.8000         1.002            High      None  

[5 rows x 25 columns]

The tail() function of pandas gives the bottom few rows of the dataset.

In [138]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Postal Code', 'City',
       'State', 'Country', 'Region', 'Market', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit', 'Shipping Cost', 'Order Priority', 'Returned'],
      dtype='object')

The columns function of pandas gives the names of all the columns present in the dataset without missing any columns.

In [139]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          51290 non-null  int64         
 1   Order ID        51290 non-null  object        
 2   Order Date      51290 non-null  datetime64[ns]
 3   Ship Date       51290 non-null  datetime64[ns]
 4   Ship Mode       51290 non-null  object        
 5   Customer ID     51290 non-null  object        
 6   Customer Name   51290 non-null  object        
 7   Segment         51290 non-null  object        
 8   Postal Code     9994 non-null   float64       
 9   City            51290 non-null  object        
 10  State           51290 non-null  object        
 11  Country         51290 non-null  object        
 12  Region          51290 non-null  object        
 13  Market          51290 non-null  object        
 14  Product ID      51290 non-null  object        
 15  Ca

The info() function of Pandas is very helpful for the coders or the analysts as it gives the information of each and every column present in the dataset, which includes the Column Name, Count of Not Null values and the DataType of each column. 

In [140]:
df.shape

(51290, 25)

The shape function of Pandas finds out the number of rows and columns available in the dataset. It helps in knowing whether the dataset is small or large. The shape of the dataset is (51290 rows and 25 columns).

In [141]:
df.size

1282250

The size operation gives the number of values or the data present in the whole dataset. The size of the dataset is about 1282250.

### ***Finding the Missing Values***
isnull() function of pandas is been widely used in all pre-processing steps to identify the null values.

In [142]:
df.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
Postal Code       41296
City                  0
State                 0
Country               0
Region                0
Market                0
Product ID            0
Category              0
Sub-Category          0
Product Name          0
Sales                 0
Quantity              0
Discount              0
Profit                0
Shipping Cost         0
Order Priority        0
Returned          49088
dtype: int64

- ***Postal Code:*** This column gives the information abou the code of a particular region or area. This column contains many null values of about 41296. More than 75% of the column contains null values. It must be imputed from the techniques like mean, median or remove the whole column itself.

- ***Returned:*** This column represents the returned back products to the store. This column contains many null values of about 49088. More than 90% of the column contains null values. It must be imputed from the techniques like mean, median or remove the whole column itself.

- ***Other Columns:*** The remaining columns (Row ID, Order ID, Order Date, Ship Date, Ship Mode, Customer ID, Customer Name, Segment, City, State, Country, Region, Market, Product ID, Category, Sub-Category, Product Name, Sales, Quantity, Discount, Profit, Shipping Cost, Order Priority) does not contains any missing values. The analysis can be done for these columns without any imputation.

In [143]:
df.duplicated().sum()

np.int64(0)

The duplicated() function of pandas gives the number of duplicate rows present in the dataset. There are no duplicate rows or values present in the data.

#### ***Data Handling Missing Values***

In [144]:
df['Returned'].unique()

array([None, 'Yes'], dtype=object)

In [145]:
df['Returned'].fillna('No', inplace=True)

In [146]:
df.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
Postal Code       41296
City                  0
State                 0
Country               0
Region                0
Market                0
Product ID            0
Category              0
Sub-Category          0
Product Name          0
Sales                 0
Quantity              0
Discount              0
Profit                0
Shipping Cost         0
Order Priority        0
Returned              0
dtype: int64

#### ***Feature Engneering***

In [147]:
df['Year'] = pd.DatetimeIndex(df['Order Date']).year
df['Month'] = pd.DatetimeIndex(df['Order Date']).month

In [148]:
df['Discount'] = df['Discount']*100
df['Discount'] = df['Discount'].astype(int)

In [149]:
df['Total_Sales'] = df['Sales'] * df['Quantity']
df['Total_Profit'] = df['Profit'] * df['Quantity']
df['Total_Shipping_Cost'] = df['Shipping Cost'] * df['Quantity']

#### ***Data Reduction***

In [152]:
df.drop(columns = ['Postal Code', 'Row ID', 'Order ID', 'Ship Date', 'Customer ID', 'Product ID'], inplace = True)

The drop() function of pandas removes the specified rows and columns which are further not required for analysis. The columns like ***Postal Code, Returned, Row ID, Order ID and Customer ID*** are removed from the dataset which are not needed for analysis.

In [153]:
df.head(2)

Order Date     Ship Mode  Customer Name    Segment           City  \
0 2014-11-11   First Class  Aaron Bergman   Consumer  Oklahoma City   
1 2014-02-05  Second Class  Justin Ritter  Corporate     Wollongong   

             State        Country      Region        Market    Category  ...  \
0         Oklahoma  United States  Central US          USCA  Technology  ...   
1  New South Wales      Australia     Oceania  Asia Pacific   Furniture  ...   

  Discount    Profit  Shipping Cost  Order Priority  Returned  Year  Month  \
0        0   62.1544          40.77            High        No  2014     11   
1       10 -288.7650         923.63        Critical        No  2014      2   

  Total_Sales Total_Profit  Total_Shipping_Cost  
0      443.96     124.3088                81.54  
1    33384.60   -2598.8850              8312.67  

[2 rows x 24 columns]

In [154]:
df.shape

(51290, 24)